### PR TITLE
feat: Add a proving crate + benchmarks

### PR DESCRIPTION
**Pull request by huitseeker**
Adds basic split-up benchmarks with Arkworks-Groth16 on BLS12_381.

Opening early to set the crate layout.

### CODE REVIEW

File: fastcrypto-proving/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+repository = "https://github.com/MystenLabs/fastcrypto"
+
+[[bench]]
+name = "proving"
+harness = false
+
+[dependencies]
+
+[dev-dependencies]
+ark-bls12-381 = "0.3.0"
```
Comment (kchalkias) [inline fastcrypto-proving/Cargo.toml:18]: I guess this is a temp (initial) backend for benchmarking purposes, until we end up with blst backend?

File: fastcrypto-proving/benches/proving.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        let a = cs.new_witness_variable(|| self.a.ok_or(SynthesisError::AssignmentMissing))?;
+        let b = cs.new_witness_variable(|| self.b.ok_or(SynthesisError::AssignmentMissing))?;
+        let c = cs.new_input_variable(|| {
+            let a = self.a.ok_or(SynthesisError::AssignmentMissing)?;
+            let b = self.b.ok_or(SynthesisError::AssignmentMissing)?;
+
+            Ok(a * b)
+        })?;
+
+        for _ in 0..(self.num_variables - 3) {
```
Comment (kchalkias) [inline fastcrypto-proving/benches/proving.rs:36]: explain the requirement for `-3`

File: fastcrypto-proving/benches/proving.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+#[derive(Copy, Clone)]
+struct DummyCircuit<F: PrimeField> {
+    pub a: Option<F>,
+    pub b: Option<F>,
+    pub num_variables: usize,
+    pub num_constraints: usize,
+}
+
+impl<F: PrimeField> ConstraintSynthesizer<F> for DummyCircuit<F> {
+    fn generate_constraints(self, cs: ConstraintSystemRef<F>) -> Result<(), SynthesisError> {
```
Comment (kchalkias) [inline fastcrypto-proving/benches/proving.rs:26]: I'd suggest at least a simple doc comment re the dummy circuit. TL;DR does it prove that "I know two numbers with whose product is X"? (in which case a, b are witnesses and X (their product) a public input)

File: fastcrypto-proving/benches/proving.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+) {
+    static VARIABLES: [usize; 5] = [16, 32, 64, 128, 256];
+
+    for size in VARIABLES.iter() {
+        let rng = &mut ark_std::test_rng();
+        let c = DummyCircuit::<F> {
+            a: Some(<F>::rand(rng)),
+            b: Some(<F>::rand(rng)),
+            num_variables: *size,
+            num_constraints: 65536,
```
Comment (kchalkias) [inline fastcrypto-proving/benches/proving.rs:61]: how did you come up to this number?

File: fastcrypto-proving/Cargo.toml (hunk 1 above)
Comment (huitseeker) [reply fastcrypto-proving/Cargo.toml:18]: It is, though I expect we will keep arkworks around for a while, for unit-testing purposes.

File: fastcrypto-proving/benches/proving.rs (hunk 2 above)
Comment (huitseeker) [reply fastcrypto-proving/benches/proving.rs:36]: a, b, c are defined right above

File: fastcrypto-proving/benches/proving.rs (hunk 4 above)
Comment (huitseeker) [reply fastcrypto-proving/benches/proving.rs:61]: I've actually refactored the test to focus on the # of constraints, which the proof timing scales with. The number of variables makes little sense considering we only provide one input.

File: Cargo.toml [hunk 5]
```diff
@@ -2,6 +2,7 @@
 resolver = "2"
 members = [
     "fastcrypto",
-    "fastcrypto-derive"
+    "fastcrypto-derive",
+    "fastcrypto-proving"
```
Comment (kchalkias) [inline Cargo.toml:6]: I'd rename this to fastcrypto-zkproving or even better fastcrypto-zkp

**Review by kchalkias — DISMISSED** [review]
Approving, but still prefer to name the crate fastcrypto-zkp, as we plan to have fastcrypto-signature, fastcrypto-hash etc

**Review by joyqvq — APPROVED** [review]

**Pull request merged**